Sw 131/data recording UI

jrvollmer opened the pull request:
**Changes:**
- Integrated data recording into the front end.
  - Popover trigger for data recording controls can be dragged anywhere within the dashboard.
- Removed code for replaying recorded data.
- Fixed bug with recorded data being bad due to bad incoming packets.
  - Checking size of incoming packets to ensure all necessary data is included. If not, the bad packet is ignored.
- Implemented recording of solar_car_connection (status of comms with solar car).
- Updated README to reflect Python dependencies and location of recorded data.
- Ignore recorded data and list of sessions.

**Known bugs/limitations:**
- Data recording controls open when mouse is released after dragging the popover trigger.

**Notes for reviewer:**
- Please create a session, record data, and then convert that data to Excel using the dashboard. Then, run `git status` and check that there are no changes.
  - Notably, `/Backend/src/routes/*.csv`, `/Backend/recordedData/sessions/*.bin`, and `/Backend/recordedData/sessionsList.bin` should not appear as modified or new to git. However, there should be new .bin and .csv files in these locations on your local copy of the project.

jrvollmer committed:
Updated submodule and commented out replaying data

jrvollmer committed:
Data recording controls are in a popover

jrvollmer committed:
Adding sessions directory to repo

Without the sessions directory (Backend/recordedData/sessions/),
data recording doesn't work

jrvollmer committed:
Prettified and persistent popover

jrvollmer committed:
Don't track recorded data/sessions anymore

jrvollmer committed:
Moved data recording control to a separate component

jrvollmer committed:
Moved modal outside of popover

jrvollmer committed:
Data recording controls are draggable

jrvollmer committed:
Made button recording icon

jrvollmer committed:
Removed unused section

jrvollmer committed:
Close popover on blur

The popover opens when done dragging, which is annoying, especially
if the popover doesn't close on blur, so I had the popover close
on blur, and then the popover trigger button receives focus when
the create new session modal closes (otherwise, any data recording
controls within the popover maintain the popover's focus).

jrvollmer committed:
Data recording controls are bounded within the dashboard

jrvollmer committed:
Downsized data recording controls size

jrvollmer committed:
Get updated list of sessions right away

jrvollmer committed:
Only initialize sessionsList once

jrvollmer committed:
Added color mode

jrvollmer committed:
Changed 'Get recorded data' to icon

jrvollmer committed:
Added tooltips

jrvollmer committed:
Disable other controls when recording

jrvollmer committed:
Move focus to record button after choosing a session

jrvollmer committed:
Cleanup

jrvollmer committed:
Removed retrieval of recorded data

jrvollmer committed:
Changed get-recorded-data to process-recorded-data

jrvollmer committed:
Cleanup

jrvollmer committed:
More cleanup

jrvollmer committed:
Clean up files

jrvollmer committed:
Added solar car connection to recorded data

jrvollmer committed:
Fixed issue with bad file writes

jrvollmer committed:
Actually fixed bad recorded data issue

The issue was with the incomind packets, not writing data to
files. Some incoming packets had bad lengths, so adding a
check for packet lengths fixed the issues that were occurring
when processing recorded data.

jrvollmer committed:
Dynamically get number of bytes per packet

jrvollmer committed:
Remove old comments

jrvollmer committed:
Remove saved sessions

jrvollmer committed:
Keep /Backend/recordedData/sessions/

jrvollmer committed:
I don't like that solution

jrvollmer committed:
Changed .gitkeep to README

jrvollmer committed:
Removed old .bin and .txt files from Backend/src/routes

jrvollmer committed:
Updated README

jrvollmer committed:
Changed .xlsx to .csv to reflect actual file type

EUdds commented on Backend/src/routes/api.js:
```diff
@@ -100,200 +103,73 @@ ROUTER.post("/record-data", (req, res) => {
-  // TODO const python = spawn('python', ['./src/routes/exp_new_script1.py','./src/routes/demo3.bin','./Data/sc1-data-format/format.json','./src/routes/test.csv']);
-  const python = spawn('python', ['./src/routes/exp_new_script1.py',
+  // Spawn new child process to call the python script
+  const python = spawn('python', ['./src/routes/process_recorded_data.py',
```
we should check for python version here or enforce a dockerized enviornment here. python doesn't mean python for everyone else

jrvollmer replied:
Yeah, the plan is to match the docker environment. Then, I can add some instructions to the README to alias `python` (or `python3` or `py` or whatever it happens to be) to refer to an appropriate version of Python 3.X.X.

EUdds commented on Backend/src/routes/api.js:
```diff
@@ -100,200 +103,73 @@ ROUTER.post("/record-data", (req, res) => {
+  fs.appendFile("recordedData/sessions/" + currentSession + ".bin", Buffer.concat([data, Buffer.alloc(1, true)]),
+        (err) => {
+                  if(err) {
+                    console.warn("ERROR: Error while appending to file");
```
```suggestion
                    console.error("ERROR: Error while appending to file");
```

jrvollmer replied:
Addressed

EUdds commented on Backend/src/routes/api.js:
```diff
@@ -344,6 +211,23 @@ function openSocket() {
       solarCarData.solar_car_connection[0] = false;
       frontendData.solar_car_connection[0] = false;
+      // If recording, replace the latest solar_car_connection value in file with false
+      if(doRecord) {
+        fs.open("recordedData/sessions/" + currentSession + ".bin", "r+", (err, fd) => {
+          if(!err) {
+            fs.write(
+                fd, Buffer.alloc(1, false), 0, 1, fs.fstatSync(fd).size - 1,
+                (err, bw, buf) => {
+                  if(!err) {
+                    // Successfully wrote byte to offset
+                  } else {
+                    console.warn("ERROR: Error writing to file when lost connection");
+                  }
+                }
+            );
+          }
+        });
+      }
```
Lets use a if(err) exit kind of structure here. It's hard to follow. Instead of having the error condition at the bottom you should have it at the top

jrvollmer replied:
Addressed

EUdds commented on Backend/src/routes/process_recorded_data.py:
```diff
@@ -1,7 +1,6 @@
 import sys
```
apply black formatting to file. https://black.readthedocs.io/en/stable/

jrvollmer replied:
Done

EUdds commented on DataGenerator/src/server.js:
```diff
@@ -103,7 +103,7 @@ const SERVER = NET.createServer((socket) => {
 
     socket.write(buf1);
-  }, 1000); // TODO 34
+  }, 67); // TODO 67
```
What does TODO 67 mean?

jrvollmer replied:
Nothing. The TODO was the original frequency for generating random data. It's not important though, so I removed the TODO

EUdds reviewed (changes requested):
First pass done.

EUdds commented on Backend/src/routes/api.js:
```diff
@@ -1,35 +1,21 @@
-  lineNumber++;
+let bytesPerPacket = 0;
+for (const property in DATA_FORMAT) {
+  bytesPerPacket += DATA_FORMAT[property][0];
```
Change 0 to be a const labelling why it's the 0th index

jrvollmer replied:
Addressed

EUdds commented on Backend/src/routes/api.js:
```diff
@@ -39,16 +25,34 @@ ROUTER.get("/api", (req, res) => {
+// Getting all the created sessions from sessionsList.bin
+while (line = broadbandLines.next()) {
+  // console.log(`Line ${lineNumber} has: ${line.toString('ascii')}`);
+  sessionsList.push(line.toString('utf8').replace('\r', ""))
```
Let's make this a clean function

jrvollmer replied:
Addressed

EUdds commented on Backend/src/routes/api.js:
```diff
@@ -39,16 +25,34 @@ ROUTER.get("/api", (req, res) => {
+  sessionsList.push(line.toString('utf8').replace('\r', ""))
+  lineNumber++;
+}
+console.log(sessionsList)
```
Label console log

jrvollmer replied:
Addressed

EUdds commented on Backend/src/routes/api.js:
```diff
@@ -344,6 +211,23 @@ function openSocket() {
       frontendData.solar_car_connection[0] = false;
+      // If recording, replace the latest solar_car_connection value in file with false
+      if(doRecord) {
+        fs.open("recordedData/sessions/" + currentSession + ".bin", "r+", (err, fd) => {
```
Make string paths a const

jrvollmer replied:
Addressed

EUdds commented on Backend/src/routes/api.js:
```diff
@@ -344,6 +211,23 @@ function openSocket() {
+                fd, Buffer.alloc(1, false), 0, 1, fs.fstatSync(fd).size - 1,
+                (err, bw, buf) => {
+                  if(!err) {
+                    // Successfully wrote byte to offset
```
Do if err not if !err else

jrvollmer replied:
Addressed in most recent commit. Sorry for the premature comments on the previous review.

jrvollmer committed:
Changes per suggestions

Changed warnings to errors and cleaned up in api.js. Reformatted
process_recorded_data.py using Black formatting. Removed TODO in
DataGenerator's server.js.

jrvollmer committed:
More changes per suggestions

Added constants for indices and paths. Made buffer to string
conversion into a function. Labeled initial sessions list log.

EUdds commented on Backend/src/routes/process_recorded_data.py:
```diff
@@ -0,0 +1,222 @@
+import xlsxwriter
+
+
+print("#Hello from python#")
+print("First param: " + sys.argv[1])
+print("Second param: " + sys.argv[2])
```
Still necessary?

jrvollmer replied:
Nope. Removed

EUdds reviewed (approved):
one nit

jrvollmer committed:
Removed useless comments in Python script

Removed useless comments from process_recorded_data.py.
Will want to remove data recording debugging comments
before merging into main.
